### PR TITLE
Work with the new application-services release process

### PR DESCRIPTION
**Pull request by bendk**
Updated the code to handle nightly releases.  The release channel is not implemented yet, but I plan to come back and do that once the new nightly system is in place.

I tested this by merging the `app-services-nightlies` into `bendk/firefox-android/main` then running the relbot script against it.  

Here's the resulting PR: https://github.com/bendk/firefox-android/pull/3

Here's the script output:
```
2023-03-30 12:38:37,920 - __main__.<module>:106 - INFO - This is relbot working on https://github.com/bendk as sebastian@mozilla.com / MickeyMoz
2023-03-30 12:38:38,080 - util.get_current_ac_version:112 - INFO - Fetched A-C version 113.0a1 from bendk/firefox-android
2023-03-30 12:38:38,080 - android_components._update_application_services:300 - INFO - Updating A-S on bendk/firefox-android:main
2023-03-30 12:38:38,238 - android_components._update_application_services:305 - INFO - Current A-S channel is nightly
2023-03-30 12:38:38,455 - android_components._update_application_services:310 - INFO - Current A-S Nightly version in A-C bendk/firefox-android:main is 113.20230320174514
2023-03-30 12:38:38,879 - android_components._update_application_services:319 - INFO - Latest A-S Nightly version available is 113.20230323131908
2023-03-30 12:38:38,879 - android_components._update_application_services:330 - INFO - We should update A-C main with A-S Nightly 113.20230323131908
2023-03-30 12:38:39,205 - android_components._update_application_services:362 - INFO - Last commit on main is 20957faa4371be9cfd97b4f082a4f35d37e45a26
2023-03-30 12:38:39,550 - android_components._update_application_services:367 - INFO - Created branch relbot/update-as/ac-113 on 20957faa4371be9cfd97b4f082a4f35d37e45a26
2023-03-30 12:38:39,550 - android_components._update_application_services:369 - INFO - Updating android-components/plugins/dependencies/src/main/java/ApplicationServices.kt
2023-03-30 12:38:40,103 - android_components._update_application_services:386 - INFO - Creating pull request
2023-03-30 12:38:41,351 - android_components._update_application_services:393 - INFO - Pull request at https://github.com/bendk/firefox-android/pull/3
2023-03-30 12:38:41,351 - android_components._update_application_services:399 - INFO - Asking Bors to run a try build
2023-03-30 12:38:42,283 - android_components._update_geckoview:168 - INFO - Updating GeckoView on A-C bendk/firefox-android:main
2023-03-30 12:38:42,454 - android_components._update_geckoview:173 - INFO - Current GV channel is nightly
2023-03-30 12:38:42,623 - android_components._update_geckoview:178 - INFO - Current GV Nightly version in A-C bendk/firefox-android:main is 113.0.20230330094555
2023-03-30 12:38:43,476 - android_components._update_geckoview:189 - INFO - Latest GV Nightly version available is 113.0.20230330094555
2023-03-30 12:38:43,668 - android_components._update_geckoview:197 - INFO - Current Glean version in A-C bendk/firefox-android:main is 52.4.2
2023-03-30 12:38:43,771 - android_components._update_geckoview:202 - INFO - Latest bundled Glean version available is 52.4.2
2023-03-30 12:38:43,771 - android_components._update_geckoview:205 - WARNING - No newer GV Nightly release found. Exiting.
```

### CODE REVIEW

**Comment by bendk** [comment]
@JohanLorenzo Do you think this code could work?   I'm hoping to get all my PRs ready before the end of the nightly cycle so we can make the swap at the start of the 114 cycle.

File: src/util.py [hunk 1]
```diff
@@ -338,36 +389,27 @@ def as_version_sort_key(v):
...
+        r = requests.get(
+            taskcluster_indexed_artifact_url(
+                "project.application-services.v2.nightly.latest",
+                "public/build/nightly.json",
+            )
+        )
+        r.raise_for_status()
+        return r.json()["version"]
+    else:
+        raise NotImplementedError("Only the AS nightly channel is currently supported")
```
Comment (JohanLorenzo) [inline src/util.py:405]: Ah, relbot doesn't "ride the train". Any change we make and release has to be compatible with nightly, beta, and release at once. I'm sorry about this. The good news is: this logic will go away once [bug android-to-gecko-repos](https://bugzilla.mozilla.org/show_bug.cgi?id=android-to-gecko-repos) is done

**Review by JohanLorenzo — COMMENTED** [review]
The overall logic looks good to me. Thanks for taking care of these changes!

If I'm not mistaken, we'll have to support the old `get_latest_as_version()` logic for beta and release. Otherwise, we will break these branches. It should be a matter of having this logic: 

```python
if as_major_version >= CUTOFF_VERSION:
   # new logic
else:
  # old logic
```

**Comment by bendk** [comment]
> The overall logic looks good to me. Thanks for taking care of these changes!
> 
> If I'm not mistaken, we'll have to support the old `get_latest_as_version()` logic for beta and release. Otherwise, we will break these branches. It should be a matter of having this logic:

Ahh, I see.  I updated the code to handle both cases.  It turned out that it was usually easier to check the a-c version than the a-s one.  I manually triggered PRs for both the new and legacy case on my firefox-android repo and they looked good to me.

File: src/util.py [hunk 2]
```diff
@@ -283,29 +300,97 @@ def get_recent_fenix_versions(repo):
...
+    if match := regex.search(src):
+        return validate_as_version(match[1])
+    raise Exception(
+        f"Could not find application services version in {os.path.basename(path)}"
     )
-    return match_as_version(content_file.decoded_content.decode("utf8"))
+
+
+def match_as_channel(src):
+    """Find the GeckoView channel in the contents of the given Gecko.kt file."""
```
Comment (jcristau) [inline src/util.py:366]: this doc string looks wrong?

File: src/android_components.py [hunk 3]
```diff
@@ -353,7 +373,7 @@ def _update_application_services(
 
         log.info(
             "Updating android-components/plugins/dependencies/src"
-            "/main/java/DependenciesPlugin.kt"
+            "/main/java/ApplicationServices.kt"
```
[the remark this answers is not in the document]
Comment (bendk) [reply src/android_components.py]: Good point, I moved the logging into that function.

**Review by jcristau — APPROVED** [review]

**Review by JohanLorenzo — COMMENTED** [review]
Thank you for taking care of the review, @jcristau! Sorry, I missed that one, 2 weeks ago.

**Comment by jcristau** [comment]
@bendk firefox-android pins a specific tag of relbot (https://github.com/mozilla-mobile/firefox-android/blob/c1add9b7278852b5d80a5a3750a6d811d928828f/.github/workflows/ac-update-geckoview.yml#L32), so it shouldn't matter when this gets merged, we can make a relbot release and then include a bump of the relbot version in the firefox-android PR.

**Comment by bendk** [comment]
Ahh, very nice.

Could you make a release with this change?  Once that's done, I'll update my firefox-android PR to include the bump.

**Comment by jcristau** [comment]
Done: https://github.com/mozilla-mobile/relbot/releases/tag/6.1.0

**Pull request merged**